contracts/tokens/extension/ExtendableDiamond.sol

```diff
@@ -0,0 +1,260 @@
+import {ExtensionRegistrationStorage} from "./ExtensionRegistrationStorage.sol";
+import {LibDiamond} from "../../diamond/libraries/LibDiamond.sol";
+import {IDiamondCut} from "../../diamond/interfaces/IDiamondCut.sol";
+import {Diamond} from "../../diamond/Diamond.sol";
+import {IToken, TokenStandard} from "../IToken.sol";
+
+/**
+* @title Router contract for Extensions
+* @notice This should be inherited by token proxy contracts
+* @dev ExtendableDiamond provides internal functions to manage
+* extensions, view extension data and invoke extension functions
+* (if the current call is an extension function) through the Diamond EIP
+*/
+contract ExtendableDiamond is TokenEventManagerStorage, ExtensionRegistrationStorage, Diamond {
+
+    /**
+    * @dev Register an extension at the given global extension address. This will create a new
+    * DiamondCut with the extension address being the facet. All external functions the extension
+    * exposes will be registered with the DiamondCut. The DiamondCut will be initalized by calling
+    * the initialize function on the extension through delegatecall
+    * Registering an extension automatically enables it for use.
+    *
+    * @param extension The global extension address to register as a Diamond facet
+    */
+    function _registerExtension(address extension) internal virtual returns (bool) {
+        MappedExtensions storage extLibStorage = _extensionStorage();
+        require(extLibStorage.extensions[extension].state == ExtensionState.EXTENSION_NOT_EXISTS, "The extension must not already exist");
+
+        //Interfaces has been validated, lets begin setup
+
+        IExtension ext = IExtension(extension);
+
+        //Next lets figure out what external functions to register in the Extension
+        bytes4[] memory externalFunctions = ext.externalFunctions();
+
+        IDiamondCut.FacetCut[] memory _diamondCut = new IDiamondCut.FacetCut[](1);
+        _diamondCut[0] = IDiamondCut.FacetCut(
+            extension,
+            IDiamondCut.FacetCutAction.Add,
+            externalFunctions
+        );
+        bytes memory initCalldata = abi.encodePacked(abi.encodeWithSelector(IExtension.initialize.selector), extension);
+        LibDiamond.diamondCut(_diamondCut, extension, initCalldata);
+
+        //Finally, add it to storage
+        extLibStorage.extensions[extension] = ExtensionData(
+            ExtensionState.EXTENSION_ENABLED,
+            extLibStorage.registeredExtensions.length,
+            address(this),
+            externalFunctions
+        );
+
+        extLibStorage.registeredExtensions.push(extension);
+        return true;
+    }
+
+    function _upgradeExtension(address extension, address newExtension) internal returns (bool) {
+        MappedExtensions storage extLibStorage = _extensionStorage();
+        require(extLibStorage.extensions[extension].state != ExtensionState.EXTENSION_NOT_EXISTS, "The extension must already exist");
```

Comment:
When we refer to an extension being enabled or disabled, do we mean that purely in regard to usage? Do we care if an extension is disabled if we try and upgrade it?